frontend_tests/node_tests/settings_user_groups.js

```diff
@@ -268,6 +289,163 @@ set_global('people', {
+        return noop;
+    };
+
+    $(".organization").on = function () {
+        return noop;
+    };
+
+    var user_group_find_called = 0;
+    var turned_off = {};
+    var on_click_called = false;
+    var pill_container_find_called = 0;
+    var pill_hover_called = false;
+    var exit_button_called=false;
+    var callback;
+    var user_group_stub = $('div.user-group[id="1"]');
+    var name_field_stub = $.create('fake-name-field');
+    var description_field_stub = $.create('fake-description-field');
+    var input_stub = $.create('fake-input');
+    user_group_stub.find = function (elem) {
+        if (elem === '.name') {
+            user_group_find_called += 1;
+            return name_field_stub;
+        }
+        if (elem === '.description') {
+            user_group_find_called += 1;
+            return description_field_stub;
+        }
+    };
+
+    var pill_container_stub = $('.pill-container[data-group-pills="Mobile"]');
+    // Test the 'off' handlers on the pill-container
+
+    pill_container_stub.off = function (event_name, sel) {
+        if (sel === undefined) {
+           sel = 'whole';
+        }
+        turned_off[event_name + '/' + sel] = true;
+    };
+
+    pill_container_stub.on = function (event_name, handler) {
+        var event = {
+            stopPropagation: noop,
+        };
+        assert.equal(event_name, 'click');
+        on_click_called = true;
+        handler(event);
+    };
+
+    var pill_stub = $.create('fake-pill');
+
+    pill_container_stub.find = function (elem) {
+        if (elem === '.input') {
+            pill_container_find_called += 1;
+            return input_stub;
+        }
+        if (elem === '.pill') {
+            pill_container_find_called += 1;
+            return pill_stub;
+        }
+    };
+
+    pill_stub.hover = function (one, two) {
+        callback = one;
+        pill_hover_called = true;
+        assert.equal(typeof(one), 'function');
+        assert.equal(typeof(two), 'function');
+    };
+
+    var exit_button = $.create('fake-pill-exit');
+    pill_stub.set_find_results('.exit', exit_button);
+
+    exit_button.css = function (property, value) {
+        exit_button_called = true;
+        assert.equal(property, 'opacity');
+        assert.equal(value, '0.5');
+    };
+
+    // Test click handlers with an external user
+    $('#user-groups #1').on = function (event_name, child_selector, fn) {
+        assert.equal(fn(), undefined);
+    };
+
+    $('#user-groups').on = function (event_name, child_selector, fn) {
+        if (event_name === 'click' && child_selector === '.delete') {
+            var fake_delete = $.create('fk-#user-groups.delete_btn');
+            fake_delete.set_parents_result('.user-group', $('.user-group'));
+            $('.user-group').attr('id', '1');
+            assert.equal(fn.call(fake_delete), undefined);
+        }
+        if (event_name === 'keypress' && child_selector === '.user-group h4 > span') {
```

Comment:
You don't really need the `keypress` section here.  I tried removing it and the tests still pass.  (You don't need to return a `noop` function, because nothing is going to execute this anyway.)

Another thing to note is that `zjquery` actually has a default `.on(...)` function that you can probably just let run.  And then when you want to test the individual handlers you can do something like this:

~~~
settings_user_group.set_up();
delete_handler = $('#user_groups').get_on_handler('.click', '.delete.');
delete_handler();

Reply:
@showell since settings_user_groups.set_up() is being called twice in whole testing, due to which 'on' is being set twice due to which get_on_handler is working only for first time and is interfering in tests of test_on_events(), so I think it's better to have custom 'on' method in my case, to prevent conflicting tests.
Regarding keypress part, I'll remove that.

Reply:
You should just reset all of the simulated jQuery before doing `set_up()` the second time.  This way you'll prevent all possible leaks.  If you need to duplicate some setup, just make functions for those.

~~~
set_global('$', global.make_zjquery());

Reply:
@showell  I found a shorter method, instead of resetting zjquery, since I just need to remove duplicate 'on' functions, isn't it better that I keep the 'on' click handlers when set_up() was first called, and when I call it the second time, I still use get_on_handlers, but just simulate the 'on' to return nothing, in turn preventing it register duplicate 'on' handlers. I think it faster and easy to implement and will reduce writing duplicate code. Let me know what you think!

Reply:
I would just reset zjquery every time you call `set_up`.  Resetting zjquery isn't expensive at all, and leaks are a pain to debug.

Reply:
I understand your objection here, but I already addressed it two comments ago.  If there is setup that you need to repeat both times you do setup, you can de-duplicate this by creating functions.You can create a function named something like `reset_test_setup` that does the following:

* reset zjquery
* call `settings_user_groups.set_up`
* build up some fake DOM elements

Basically, all the common stuff can be put in a function.  All the non-common stuff needs to be non-common, anyway.

Reply:
So, then call this function, every time before doing set_up() ?

Reply:
Thanks!  The function wrappers make the subsequent diff a lot more readable!